### PR TITLE
Changes needed when using the TestGrid local deployment

### PR DESCRIPTION
**Pull request by udhanMti**
**Purpose**

Package all components of Testgrid - TG-Java, Jenkins, Dashboard, Database(H2) and then make it executable on local Linux instances.
Blank Diagram(1)

**Sample User Story**

1.  Get the localTestgrid zip file and unzip it

2.  Move into the unzipped folder and enter the credentials in credentials.config file

3.  Move into the testgrid-home folder and enter configuration properties in configurations.properties file

4.  Open the terminal within the unzipped folder and execute command ./local-tg startup-jenkins

     (Then Tomcat server will be started and all the plugins needed will be installed in Jenkins engine. The very first time it will take about 10 minutes. Then the credentials will be added to Jenkins engine and H2 server will be started )

5.  Access Jenkins engine at localhost:8080/jenkins URL

     (User will see testgrid-job-creator job listed in there)

6.  When user build the testgrid-job-creator job all other jobs currently in the https://testgrid-live-dev.private.wso2.com will be listed in the Jenkins dashboard

7.  Then the user can build any of the jobs listed in the Jenkins dashboard.

8.  Execute command ./local-tg shutdown-jenkins to shutdown Tomcat server and H2 server.

### CODE REVIEW

File: deployment-local/jenkins-home/config.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      <filterQueue>false</filterQueue>
+      <properties class="hudson.model.View$PropertyList"/>
+    </hudson.model.AllView>
+  </views>
+  <primaryView>all</primaryView>
+  <slaveAgentPort>0</slaveAgentPort>
+  <label></label>
+  <nodeProperties/>
+  <globalNodeProperties/>
+</hudson>
```
Comment (chirangaalwis) [inline deployment-local/jenkins-home/config.xml]: @udhanMti please add a new EOF line. Applies to all similar cases within this PR.

File: deployment-local/add_credentials.sh [hunk 2]
```diff
@@ -0,0 +1,185 @@
+#!/bin/bash
```
Comment (kasunbg) [inline deployment-local/add_credentials.sh]: Please move all these resources in to deployment-local/src/main/resources folder.

File: deployment-local/pom.xml [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <parent>
+        <groupId>org.wso2.testgrid</groupId>
+        <artifactId>testgrid-parent</artifactId>
+        <version>1.0.8-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>org.wso2.testgrid.deployment.local</artifactId>
```
Comment (kasunbg) [inline deployment-local/pom.xml]: We can keep this as the artifact id.

But, we need a better name for the actual zip distirbution. Say, something like testgrid-local-{version}.zip.

File: deployment-local/pom.xml [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    <name>WSO2 TestGrid Project - Deployment Local POM</name>
```
Comment (kasunbg) [inline deployment-local/pom.xml]: When I unzip the distribution, I saw multiple different files directly under root folder location. This is quite confusing for a new user. Hence, please move files users does not interact with to a separate folder.

File: deployment-local/local-tg [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+     ./tomcat/apache-tomcat-8.5.43/bin/shutdown.sh
+   fi
+
+   if [ $(curl -s -o /dev/null -w "%{http_code}" http://localhost:8082) -eq 200 ]
+   then
+     java -cp lib/h2-*.jar org.h2.tools.Server -tcpShutdown tcp://localhost:9092
+   fi
+
+else
+   echo "Invalid command"
```
Comment (kasunbg) [inline deployment-local/local-tg]: There's an issue here in terms of user experience.

As a first time user, I ran the ./local-tg command. It has simply output "Invalid command", and exited. I was left wondering what to do next.
image

Please help text such that user will know what to do next. To know how a command-line should behave, analyze the log output of the wso2 'wum' client. Here's a sample screenshot for reference.

image

File: deployment-local/local-tg [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#!/bin/sh
+
+if [ "$1" = "startup-jenkins" ]; then
```
Comment (kasunbg) [inline deployment-local/local-tg]: What does the user expect to see here? When I ran this command, following log was output. As a first-time user, this kept me wondering what I was supposed to do next.

image

File: deployment-local/local-tg [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    echo "__          _______  ____ ___    _______        _    _____      _     _    _       ____   _____   ___   _      "
+    echo "\ \        / / ____|/ __ \__ \  |__   __|      | |  / ____|    (_)   | |  | |    /  __  \/ _____|/    \| |     "
```
Comment (kasunbg) [inline deployment-local/local-tg]: The startup-jenkins was stuck in the middle without any errors/log output. We can improve this.

Comment (kasunbg) [reply deployment-local/local-tg]: I was supposed to set values for following two properties in params.properties. After setting that, server started fine. So, what's missing was the initial user experience.

```
uat_nexus_settings_path=uat-nexus.xml
common_configs_path=common-configs.properties
```

File: deployment-local/local-tg [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+
+    if [ $(curl -s -o /dev/null -w "%{http_code}" http://localhost:8082) -eq 000 ]
+    then
+        echo "starting H2 server....."
+        java -jar lib/h2-*.jar  -baseDir ./h2dbs &> /dev/null &
+    fi
+
+
+elif [ "$1" = "shutdown-jenkins" ]
```
Comment (kasunbg) [inline deployment-local/local-tg]: If user pressed ctrl+c / send a SIGTERM/SIGKILL signal via kill command, then we should shutdown testgrid. So, we should execute the logic under this.

You can Linux's TRAP command for this purpose.

File: deployment-local/local-tg (hunk 7 above)
Comment (udhanMti) [reply deployment-local/local-tg]: I have added a log output.

**Pull request merged**